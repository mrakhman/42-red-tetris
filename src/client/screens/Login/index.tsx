import { StackNavigationProp } from '@react-navigation/stack';
import { StyleSheet, Text, TouchableOpacity, TextInput, View } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import * as React from 'react';

import Gameboy from '/client/components/Gameboy';

export default function Login(): JSX.Element {
  const navigation = useNavigation<StackNavigationProp<RootStackParamList, 'Root'>>();
  const [username, setUsername] = React.useState<string | null>();
  const [room, setRoom] = React.useState<string | null>();
  return (
    <Gameboy>
      <View style={{ alignItems: 'center' }}>
        <Text style={styles.title}>{'Username'}</Text>
        <TextInput
          value={username ?? ''}
          onChangeText={setUsername}
          style={{ borderWidth: 1, marginBottom: 20, height: 30 }}
        />
        <Text style={styles.title}>{'Room'}</Text>
        <TextInput
          value={room ?? ''}
          onChangeText={setRoom}
          style={{ borderWidth: 1, marginBottom: 20, height: 30 }}
        />
        <TouchableOpacity
          style={{ marginBottom: 20 }}
          onPress={() => username && room && navigation.push('Playground', { username, room })}
        >
          <Text style={styles.linkText}>Join Room</Text>
        </TouchableOpacity>
        <TouchableOpacity
          onPress={() => navigation.push('Playground', {})}
        >
          <Text style={styles.linkText}>Solo mode</Text>
        </TouchableOpacity>
      </View>
    </Gameboy>
  );
}

const styles = StyleSheet.create({
  title: {
    marginBottom: 10,
    alignSelf: 'flex-start',
  },
  linkText: {
    fontSize: 14,
    color: '#004580',
  },
});
