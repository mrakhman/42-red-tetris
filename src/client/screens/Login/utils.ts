import axios from 'axios';

import { maxPlayersLimit } from '/client/constants/game';

export const isEmpty = (value?: string | null): boolean => value === undefined || value === null || value === '';

export const checkTextLength = (text?: string | null): boolean => {
  const maxLength = 15;
  if (isEmpty(text)) return false;
  if (text && text.length > maxLength) return false;
  return true;
};

export const checkSpecialChars = (text?: string | null): boolean => {
  const regexpr =  new RegExp('\\W'); // matches any non-word-character, to only allow [a-zA-Z0-9_]
  if (!text) return false;
  if (text.match(regexpr)) return false;
  return true;
};

export const checkUsername = async (username: string | null | undefined): Promise<boolean> => {
  if (!checkTextLength(username)) throw Error('Name must be 1-15 symbols');
  if (!checkSpecialChars(username)) throw Error('Use letters and numbers');

  const response = await axios.get(`${process.env.SERVER_URL}/player/${username}`);
  if (response && isEmpty(response.data.username)) return(true);
  throw Error('Username already taken');
};

export const checkRoomName = async (roomName: string | null | undefined): Promise<boolean | undefined> => {
  if (!checkTextLength(roomName)) throw Error('Name must be 1-15 symbols');
  if (!checkSpecialChars(roomName)) throw Error('Use letters and numbers');

  const response = await axios.get(`${process.env.SERVER_URL}/room/${roomName}`);
  if (response && isEmpty(response.data.name)) return true;
  throw Error('Room name already taken');
};


export const isRoomPlayersLimitAvailable = async (roomName: string | null | undefined): Promise<boolean | undefined> => {
  const response = await axios.get(`${process.env.SERVER_URL}/room/${roomName}`);
  if (response && response.data.players.length >= maxPlayersLimit)
    throw Error('Too many players in the room');
  return true;
};
