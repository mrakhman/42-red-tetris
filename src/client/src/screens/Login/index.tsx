import { StackScreenProps } from '@react-navigation/stack';
import * as React from 'react';
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native';

export default function Login({
  navigation,
}: StackScreenProps<RootStackParamList, 'Root'>): JSX.Element {
  return (
    <View style={styles.container}>
      <Text style={styles.title}>{'Login'}</Text>
      <TouchableOpacity onPress={() => navigation.push('Room')} style={styles.link}>
        <Text style={styles.linkText}>Go to room!</Text>
      </TouchableOpacity>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 20,
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  link: {
    marginTop: 15,
    paddingVertical: 15,
  },
  linkText: {
    fontSize: 14,
    color: '#2e78b7',
  },
});
