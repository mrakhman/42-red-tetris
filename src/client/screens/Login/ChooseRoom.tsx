import  React, { useState, useEffect, useContext, Dispatch, SetStateAction } from 'react';
import { StackNavigationProp } from '@react-navigation/stack';
import { StyleSheet, Text, TouchableOpacity, TextInput, View } from 'react-native';
import { useNavigation } from '@react-navigation/native';

import SocketContext from '/client/context/SocketContext';
import { checkTextLength } from '/client/screens/Login/utils';

type Props = {
	username: string | null | undefined,
	roomName: string | null | undefined,
	setRoomName: Dispatch<SetStateAction<string | null | undefined>>,
};

export default function ChooseRoom(props: Props): JSX.Element {
	const {username, roomName, setRoomName} = props;

  const socket = useContext(SocketContext);
  const navigation = useNavigation<StackNavigationProp<RootStackParamList, 'Root'>>();
  const [roomNameError, setRoomNameError] = useState<string>('');
  const [waitingRooms, setWaitingRooms] = useState<string[]>([]);

  console.log('Updates for waitingRooms:', waitingRooms);
  useEffect(() => {
    if (!socket) throw Error('No socket');
    // If I remove [] from useEffect there is a flood of sockets, for some reason component rerenders several times a second.
    // Later we need to listen to a socket event when game in room has started.
    socket.emit('fetch waiting rooms');
    socket.on('all waiting rooms', (roomNames: string[]) => {
      setWaitingRooms(roomNames);
    });
  }, []);
  
  useEffect(() => {
    if (!socket) throw Error('No socket');
    // new room is created + room is deleted
    socket.on('update waiting rooms', (roomNames: string[]) => {
      setWaitingRooms(roomNames);
    });
  });

  const onJoinRoomPress = () => {
    if(!checkTextLength(roomName)) setRoomNameError('Name must be 1-15 symbols');
		else
			username && roomName && navigation.push('Playground', { username, room: roomName });
  };

  return (
    <View style={{ width: '100%', alignItems: 'center' }}>
      <Text style={styles.title}>Create room</Text>
      <Text style={styles.error}>{roomNameError}</Text>
      <TextInput
        value={roomName ?? ''}
        onChangeText={text => {
          setRoomNameError('');
          setRoomName(text);
        }}
        style={{ borderWidth: 1, marginBottom: 20, height: 30, width: '100%' }}
      />
			<TouchableOpacity
        style={styles.button}
        onPress={onJoinRoomPress}
      >
        <Text style={styles.linkText}>Play</Text>
      </TouchableOpacity>
			<View style={{ width: '80%', marginTop: 20 }}>
        {waitingRooms.length > 0 && <Text style={styles.title}>Join room</Text>}
        {waitingRooms.map((waitingRoom) =>
          <View key={waitingRoom}>
            <TouchableOpacity
              style={styles.roomsList}
              onPress={() => waitingRoom && username && navigation.push('Playground', { username, room: waitingRoom })}
						>
              <Text>{waitingRoom}</Text>
            </TouchableOpacity>
          </View>)}
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  title: {
    marginBottom: 5,
    alignSelf: 'flex-start',
    textAlign: 'center',
  },
  error: {
    marginBottom: 5,
    alignSelf: 'flex-start',
    textAlign: 'center',
    color: '#980f0f',
  },
  linkText: {
    fontSize: 14,
	},
	roomsList: {
		borderWidth: 1,
		borderRadius: 5,
		margin: 3,
		padding: 2,
		alignItems:
		'center',
	},
	button: {
		width: '70%',
		borderWidth: 1,
		borderRadius: 10,
		alignItems: 'center',
		padding: 3,
		marginBottom: 10,
	},
});
