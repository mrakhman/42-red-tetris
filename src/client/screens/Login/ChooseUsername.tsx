import  React, { useState, useContext, Dispatch, SetStateAction } from 'react';
import { StackNavigationProp } from '@react-navigation/stack';
import { StyleSheet, Text, TouchableOpacity, TextInput, View } from 'react-native';
import { useNavigation } from '@react-navigation/native';

import SocketContext from '/client/context/SocketContext';
import { checkUsername } from '/client/screens/Login/utils';

type Props = {
	username: string | null | undefined,
	setUsername: Dispatch<SetStateAction<string | null | undefined>>,
	setScreenNumber: Dispatch<SetStateAction<1 | 2>>,
};

export default function ChooseUsername(props: Props): JSX.Element {
  const {username, setUsername, setScreenNumber} = props;
  
  const socket = useContext(SocketContext);
  const navigation = useNavigation<StackNavigationProp<RootStackParamList, 'Root'>>();
  const [ussernameError, setUsernameError] = useState<string>('');

  const validateUsername = () => {
    checkUsername(username)
      .then(data => {
        if (data === true) {
          if(!socket) throw Error('No socket');
          socket.emit('create user', username);
          setScreenNumber(2);
        }
      })
      .catch(err => {
        setUsernameError(err.message);
      });
  };

  return (
    <View style={{ width: '100%', alignItems: 'center' }}>
      <Text style={styles.title}>{'Username'}</Text>
      <Text style={styles.error}>{ussernameError}</Text>
      <TextInput
        value={username ?? ''}
        onChangeText={text => {
          setUsernameError('');
          setUsername(text);
        }}
        style={{ borderWidth: 1, marginBottom: 20, height: 30, width: '100%' }}
      />
      <TouchableOpacity
        style={styles.button}
        onPress={validateUsername}
      >
        <Text style={styles.linkText}>Next</Text>
      </TouchableOpacity>
      <TouchableOpacity
        style={styles.button}
        onPress={() => navigation.push('Playground', {})} // TODO: do we need to validate username here too?
      >
        <Text style={styles.linkText}>Solo mode</Text>
      </TouchableOpacity>
    </View>
  );
}

const styles = StyleSheet.create({
  title: {
    marginBottom: 5,
    alignSelf: 'flex-start',
    textAlign: 'center',
  },
  error: {
    marginBottom: 5,
    alignSelf: 'flex-start',
    textAlign: 'center',
    color: '#980f0f',
  },
  linkText: {
    fontSize: 14,
  },
  button: {
    width: '70%',
    borderWidth: 1,
    borderRadius: 10,
    alignItems: 'center',
    padding: 3,
    marginBottom: 30,
  },
});
