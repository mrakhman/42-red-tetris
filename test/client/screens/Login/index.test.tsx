import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import ChooseUsername from '/client/screens/Login/ChooseUsername';
import ChooseRoom from '/client/screens/Login/ChooseRoom';
import Login from '/client/screens/Login/index';

describe('Snapshots', () => {
  let props: JSX.IntrinsicAttributes & { setScreenNumber: React.Dispatch<React.SetStateAction<1 | 2>>; };

  beforeEach(() => {
    props = {
      setScreenNumber: jest.fn(() => 1),
    };
  });
  it('should render ChooseUsername', () => {
    const wrapper = shallow(<ChooseUsername {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render ChooseRoom', () => {
    const wrapper = shallow(<ChooseRoom {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render Login', () => {
    const wrapper = shallow(<Login/>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
