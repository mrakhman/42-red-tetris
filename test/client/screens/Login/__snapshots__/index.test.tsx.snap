// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots should render ChooseRoom 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 16,
        "marginBottom": 50,
      }
    }
  >
    Hello, 
  </Text>
  <Text
    style={
      Object {
        "alignSelf": "flex-start",
        "marginBottom": 5,
        "textAlign": "center",
      }
    }
  >
    Create room
  </Text>
  <Text
    style={
      Object {
        "alignSelf": "flex-start",
        "color": "#980f0f",
        "marginBottom": 5,
        "textAlign": "center",
      }
    }
  />
  <Component
    allowFontScaling={true}
    onChangeText={[Function]}
    rejectResponderTermination={true}
    style={
      Object {
        "borderWidth": 1,
        "height": 30,
        "marginBottom": 20,
        "width": "100%",
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <ForwardRef
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 10,
        "borderWidth": 1,
        "marginBottom": 10,
        "padding": 3,
        "width": "70%",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 14,
        }
      }
    >
      Play
    </Text>
  </ForwardRef>
  <View
    style={
      Object {
        "marginTop": 20,
        "width": "80%",
      }
    }
  />
</View>
`;

exports[`Snapshots should render ChooseUsername 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "alignSelf": "flex-start",
        "marginBottom": 5,
        "textAlign": "center",
      }
    }
  >
    Username
  </Text>
  <Text
    style={
      Object {
        "alignSelf": "flex-start",
        "color": "#980f0f",
        "marginBottom": 5,
        "textAlign": "center",
      }
    }
  />
  <Component
    allowFontScaling={true}
    onChangeText={[Function]}
    rejectResponderTermination={true}
    style={
      Object {
        "borderWidth": 1,
        "height": 30,
        "marginBottom": 20,
        "width": "100%",
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <ForwardRef
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 10,
        "borderWidth": 1,
        "marginBottom": 30,
        "padding": 3,
        "width": "70%",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 14,
        }
      }
    >
      Next
    </Text>
  </ForwardRef>
  <ForwardRef
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 10,
        "borderWidth": 1,
        "marginBottom": 30,
        "padding": 3,
        "width": "70%",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 14,
        }
      }
    >
      Solo mode
    </Text>
  </ForwardRef>
</View>
`;

exports[`Snapshots should render Login 1`] = `
<Gameboy>
  <View
    style={
      Object {
        "alignItems": "center",
        "width": "60%",
      }
    }
  >
    <ChooseUsername
      setScreenNumber={[Function]}
    />
  </View>
</Gameboy>
`;
